diff --git a/libm/arm/e_pow.S b/libm/arm/e_pow.S
index a984d4c..55b9dab 100644
--- a/libm/arm/e_pow.S
+++ b/libm/arm/e_pow.S
@@ -429,4 +429,4 @@ ENTRY(pow_neon)
 
 .Ltwoto1o4: @ 2^1/4
     .long       0x0a31b715, 0x3ff306fe
-END(pow)
+END(pow_neon)
diff --git a/libm/src/e_pow.c b/libm/src/e_pow.c
index 50b07a6..fb3ba07 100644
--- a/libm/src/e_pow.c
+++ b/libm/src/e_pow.c
@@ -13,7 +13,7 @@
 static char rcsid[] = "$FreeBSD: src/lib/msun/src/e_pow.c,v 1.11 2005/02/04 18:26:06 das Exp $";
 #endif

-#if defined(KRAIT_NEON_OPTIMIZATION)
+#if ( defined(KRAIT_NEON_OPTIMIZATION) && !defined(NO_POW_NEON) )
 double pow_neon(double x, double y);
 #endif

@@ -205,7 +205,7 @@ __ieee754_pow(double x, double y)
            t1 = u+v;
            SET_LOW_WORD(t1,0);
            t2 = v-(t1-u);
-#if defined(KRAIT_NEON_OPTIMIZATION)
+#if ( defined(KRAIT_NEON_OPTIMIZATION) && !defined(NO_POW_NEON) )
        } else if (ix <= 0x40100000 && iy <= 0x40100000 && hy > 0 && hx > 0) {
                return pow_neon(x,y);
 #endif
